keps/sig-cli/5295-kyaml/README.md

```diff
@@ -0,0 +1,1428 @@
+
+### Non-Goals
+
+1. Introduce alternative configuration languages that are not compatible with existing tooling.
+2. Server-side support for KYAML distinct from what we support in YAML already.
+3. Require projects to migrate to KYAML by default.
+4. Change the YAML library used by Kubernetes.
+
+## Proposal
+
+KYAML is primarily a specification of output.
+Since KYAML is a subset of YAML,
+it should always be accepted as input,
+but this KEP does not demand that such input be strict KYAML.
+Users can choose to violate this specification,
+As long as the result is still valid YAML.
+
+KYAML addresses the problems cited above by:
+1. Always quoting _value_ strings (no ambiguity around things like "NO").
+2. Allowing (but not requiring) quoted keys, but always quoting them if they are not obviously safe (e.g. "no" would always be quoted).
+3. Always using {} for structs and maps (no more obscure errors about mapping values).
+4. Always using [] for lists (no more trying to figure out if a dash changes the meaning).
+
+Taken together,
+these mean that KYAML is not whitespace-sensitive,
+which makes templating and other text-patching much easier.
+YAML calls this "flow style",
+as opposed to the conventional "block style".
+
+This might sound a lot like JSON,
+because it is!
+Remember that JSON is valid YAML.
+KYAML is friendlier than JSON for humans
+to read,
+to write,
+and to commit to source control.
+Unlike JSON, KYAML:
+1. Allows comments (when authored by users).
+2. Allows trailing commas on the last item in a list or struct/map.
+3. Does not require quoted keys.
+
+KYAML also includes a "header" (still valid YAML)
+which helps to disambiguate a KYAML document from an ill-formed JSON document
+(see: [kubernetes/kubernetes#78946](https://github.com/kubernetes/kubernetes/issues/78946)).
+
+KYAML also tries to economize on vertical space by ["cuddling"](https://blog.gskinner.com/archives/2008/11/curly_braces_to.html) some kinds of brackets together.
+
+Because KYAML is YAML, a KYAML multi-document is a YAML multi-document.
+
+To make this change, we will:
+1. Create new code (see details below) which serializes any object into KYAML.
+2. Introduce a kubectl output format, `-o=kyaml`, which uses this new package.
+3. Create tooling which reformats YAML or JSON as KYAML (see details below).
+4. Work with SIG Docs to update examples to this format, both example command-lines and examples of API objects.
+5. Look for other examples across kubernetes-affiliated git repositories, and offer PRs to convert to KYAML.
+6. Make a lot of noise about it, and try to get more people to use it
+
+The rationale for changing everything everywhere is simple.
+Many users start with examples,
+which they copy and modify.
+By changing the examples we seed the change in the ecosystem.
+One success metric might be
+that KYAML becomes the preferred format in the most common helm charts.
```

Comment:
I don't see how adopting KYAML for Helm chart templates is a tenable idea.
 
> example https://github.com/bitnami/charts/blob/main/bitnami/appsmith/templates/client/deployment.yaml

Getting the Go templating to work correctly is already challenging.
Adding more curly braces to the curly brace soup is not going to be a nice DX.
Cuddling closing YAML curlies makes it worse.

Reply:
I adapted some of that helm chart and it's (IMO) easier to read and gets rid of all the `| nindent 8` crap.

Reply:
That is about all it can get rid of while adding a lot more curlies.

But I would guess that if kyaml is used with Go templating people will still want the indentation to make `helm template ...` output readable, right?